lucet-runtime/lucet-runtime-internals/src/instance/signals.rs

```diff
@@ -333,11 +332,23 @@ extern "C" fn handle_signal(signum: c_int, siginfo_ptr: *mut siginfo_t, ucontext
     });
 
     if switch_to_host {
-        HOST_CTX.with(|host_ctx| unsafe {
-            Context::set_from_signal(&*host_ctx.get())
-                .expect("can successfully switch back to the host context");
-        });
-        unreachable!()
+        // Switch to host by preparing the context to switch when we return from the signal andler.
+        // We must return from the signal handler for POSIX reasons, so instead prepare the context
+        // that the signal handler will resume the program as if a call were made. First, by
+        // pointing the instruction pointer at `lucet_context_set`, then by preparing the argument
+        // that `lucet_context_set` should read from `rdi` - the context to switch to.
+        //
+        // NOTE: it is absolutely critical that `lucet_context_set` does not use the guest stack!
```

Comment:
full disclosure: I left most of the branch unchanged while adjusting the commits, but this `NOTE:` is new. I'd realized that there wasn't anything discussing the specific constraints on the function we continue into after handling a signal, and this addresses that.